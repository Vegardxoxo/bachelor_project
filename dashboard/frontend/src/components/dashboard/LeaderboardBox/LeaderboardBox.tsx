import {Card, CardActionArea, CardContent} from '@mui/material';
import CardOverflow from '@mui/joy/CardOverflow';
import BarChartIcon from '@mui/icons-material/BarChart';
import {useNavigate} from 'react-router-dom';
import {useRecoilValue} from 'recoil';
import React, {useEffect} from 'react';
import {userAtom} from '../../../globalVariables/variables';

interface Leaderboard {
    organization: string;
    active_percentage: number;
    rank: number;
}

/**
 * Returns a LeaderboardBox card with information on the organization's rank in the leaderboard.
 */
export function LeaderboardBox() {

    // Get the function for navigating to different routes
    const navigate = useNavigate();

    const handleCardClick = () => {
        navigate(`/leaderboard`);
    };

    // Declare and initiate state variables
    const [data, setData] = React.useState<Leaderboard[]>([]);
    const accessToken = localStorage.getItem('access');
    const userInfo = useRecoilValue(userAtom);

    /* Fetched leaderboard data from API when component mounts or access token/organization changes */
    useEffect(() => {
        const fetchData = async () => {
            const response = await fetch(
                `http://127.0.0.1:8000/api/licenses/leaderboard/?organization=${userInfo.organization}`,
                {
                    method: 'GET',
                    headers: {
                        'content-type': 'application/json',
                        authorization: `Bearer ${accessToken}`
                    }
                }
            );
            if (response) {
                const data = await response.json();
                if (response.ok) {
                    setData(data.leaderboard);
                }
            }
        };
        fetchData();
    }, []);

    /*The leaderboard card*/
    return (
        <Card
            sx={{
                width: 300,
                height: 250,
                borderRadius: 5,
                ':hover': {boxShadow: 20}
            }}
            data-testid="savingsBox">
            <CardActionArea sx={{paddingBottom: 4}} onClick={handleCardClick}>
                <CardOverflow>
                    <CardContent>
                        <div color="text.secondary" id="numbersBoxes">
                            <BarChartIcon
                                sx={{fontSize: 100, color: '#80CC9F'}}></BarChartIcon>
                            <p
                                style={{
                                    fontSize: '25px',
                                    fontWeight: 'semi-bold',
                                    marginTop: '-10px'
                                }}>
                                Din enhet er p√•{' '}
                                <span style={{fontWeight: 'bold'}}>
                  {data.map((row) => {
                      if (userInfo.organization === row.organization) {
                          return row.rank.toString();
                      } else {
                          return null;
                      }
                  })}
                                    . plass
                </span>
                            </p>
                            <p style={{fontSize: '15px', fontStyle: 'italic'}}>
                                {' '}
                                Se oversikt over topplisten med andre enheter.
                            </p>
                        </div>
                    </CardContent>
                </CardOverflow>
            </CardActionArea>
        </Card>
    );
}
