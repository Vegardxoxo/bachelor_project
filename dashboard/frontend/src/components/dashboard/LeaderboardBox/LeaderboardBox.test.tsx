import React from 'react';
import {render, waitFor} from '@testing-library/react';
import {MemoryRouter} from 'react-router-dom';
import {RecoilRoot} from 'recoil';
import {LeaderboardBox} from "./LeaderboardBox";
import renderer from "react-test-renderer";
import {act} from 'react-dom/test-utils';
import 'isomorphic-fetch';

const mockApiResponse = {
    leaderboard: [
        {
            organization: 'ExampleOrg',
            active_percentage: 50,
            rank: 1,
        },
    ],
};

jest.spyOn(global, 'fetch').mockImplementation(() =>
    Promise.resolve(new Response(JSON.stringify(mockApiResponse), {
        status: 200,
        headers: {
            'Content-Type': 'application/json',
        },
    }))
);

const fetchMock = global.fetch as jest.MockedFunction<typeof global.fetch>;

describe('LeaderboardBox', () => {
    it('renders without crashing', async () => {
        await act(async () => {
            render(
                <RecoilRoot>
                    <MemoryRouter>
                        <LeaderboardBox/>
                    </MemoryRouter>
                </RecoilRoot>
            );
        });
        await waitFor(() => expect(fetchMock).toHaveBeenCalledTimes(1));
    });

    it('matches snapshot', async () => {
        let testRenderer: renderer.ReactTestRenderer | null = null;

        await act(async () => {
            testRenderer = renderer.create(
                <RecoilRoot>
                    <MemoryRouter>
                        <LeaderboardBox/>
                    </MemoryRouter>
                </RecoilRoot>
            );
        });

        await waitFor(() => expect(fetchMock).toHaveBeenCalledTimes(2));
        expect(testRenderer!.toJSON()).toMatchSnapshot();
    });
});

// Clean up the mock implementation after tests
afterAll(() => {
    jest.restoreAllMocks();
});
