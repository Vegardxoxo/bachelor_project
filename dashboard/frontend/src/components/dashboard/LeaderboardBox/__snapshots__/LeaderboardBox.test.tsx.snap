// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaderboardBox matches snapshot 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-n97x0p-MuiPaper-root-MuiCard-root"
  data-testid="savingsBox"
>
  <button
    className="MuiButtonBase-root MuiCardActionArea-root css-uli8ma-MuiButtonBase-root-MuiCardActionArea-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <div
      className="MuiCardOverflow-root MuiCardOverflow-variantPlain MuiCardOverflow-colorNeutral css-17tq388-JoyCardOverflow-root"
    >
      <div
        className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <div
          color="text.secondary"
          id="numbersBoxes"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-9ybthn-MuiSvgIcon-root"
            data-testid="BarChartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M4 9h4v11H4zm12 4h4v7h-4zm-6-9h4v16h-4z"
            />
          </svg>
          <p
            style={
              Object {
                "fontSize": "25px",
                "fontWeight": "semi-bold",
                "marginTop": "-10px",
              }
            }
          >
            Din enhet er p√•
             
            <span
              style={
                Object {
                  "fontWeight": "bold",
                }
              }
            >
              . plass
            </span>
          </p>
          <p
            style={
              Object {
                "fontSize": "15px",
                "fontStyle": "italic",
              }
            }
          >
             
            Se oversikt over topplisten med andre enheter.
          </p>
        </div>
      </div>
    </div>
    <span
      className="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
    />
    <span
      className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
